### ISSUE TITLE
[0.8.0rc0] save_file overwriting same path returns zeros when tensors come from prior load_file

### ISSUE THREAD

**Issue by spandantiwari** [huggingface/safetensors#762]
## TL;DR

In `safetensors==0.8.0-rc.0`, calling `save_file(tensors, path)` where `tensors`
were just returned by `load_file(path)` produces a **zero-filled output file** —
both on disk and in the in-memory tensors. This appears to be a regression
vs `0.7.0` (last stable), which handles this pattern correctly.

The trigger is the path collision; the bug fires whether or not the dict is
modified between load and save.

## Repro

```python
import os, tempfile, torch
from safetensors.torch import load_file, save_file
import safetensors

print("safetensors:", safetensors.__version__)  # 0.8.0-rc.0
print("torch:", torch.__version__)              # 2.10.0

with tempfile.TemporaryDirectory() as tmp:
    path = os.path.join(tmp, "model.safetensors")

    # Write a file with non-zero data
    save_file({"w": torch.tensor([1.0, 2.0, 3.0])}, path)

    # Load it back, then save (no modification) to the SAME path
    loaded = load_file(path)
    save_file(loaded, path)

    # Re-read to inspect
    reloaded = load_file(path)
    print("on-disk after round-trip:", reloaded["w"])
    print("in-memory after save:    ", loaded["w"])
```

**Output on `0.8.0rc0`**:
```
safetensors: 0.8.0-rc.0
torch: 2.10.0
on-disk after round-trip: tensor([0., 0., 0.])
in-memory after save:     tensor([0., 0., 0.])
```

**Expected (and actual on `0.7.0`)**:
```
on-disk after round-trip: tensor([1., 2., 3.])
in-memory after save:     tensor([1., 2., 3.])
```

## Variant matrix

I tested seven variations to characterize the trigger:

| # | Pattern | 0.8.0rc0 result |
|---|---|---|
| 1 | load → rename keys → save same path | **fails (zeros)** |
| 2 | load → rename keys → save **different** path | passes |
| 3 | load → rename + `.clone()` → save same path | passes |
| 4 | load → save same path, **no modification** | **fails (zeros)** |
| 5 | load → in-place `tensor.add_(100)` → save same path | **fails (zeros)** |
| 6 | load → rename → `del` + `gc.collect()` → save same path | **fails (zeros)** |
| 7 | load → rename + `.detach().clone()` → save same path | passes |

So:
- The bug is **independent of whether the dict was modified** (variants 1, 4, 5 all fail).
- It is triggered **only when the save target is the same path** as the prior load (variant 2 passes).
- It is avoided by **materializing the tensors via `.clone()`** before save (variants 3, 7 pass).
- A `del` + `gc.collect()` is **not sufficient** to release the mmap before save (variant 6 fails).

## Likely cause (speculation)

`load_file` returns memory-mapped tensors backed by the input file. When
`save_file` is then called on the same path, the writer evidently truncates the
file before serializing, which zeroes the underlying mmap pages. The serializer
then reads from the (now-zero) tensors and writes zeros back to disk. The
in-memory tensors are also zero because they're mmap views of the same file.

If this analysis is correct, the fix is one of:
- Detect the path collision and either copy first, or refuse with a clear error
- Always read each input tensor into a CPU buffer before truncating the output

## Workarounds (for downstream users hitting this now)

1. **Save to a different path** (and rename atomically afterwards if needed).
2. **Clone every tensor** before passing to `save_file`:
   ```python
   save_file({k: v.clone() for k, v in loaded.items()}, path)
   ```

## Environment

- safetensors 0.8.0-rc.0 (installed via `pip install --pre safetensors==0.8.0rc0`)
- Python 3.13
- torch 2.10.0+rocm7.1 (also reproducible with CPU-only torch — the bug is
  not GPU-related)
- Linux 6.8.0 / Ubuntu 22.04
- Hardware: AMD MI300X (irrelevant — bug is in CPU file I/O)

Confirmed **does not** reproduce on `safetensors==0.7.0` with the same Python /
torch versions.

## Question

Is a fix already in `main` past `0.8.0rc0`? If so, happy to retest against the
current dev build. Otherwise, would a same-path-detection + auto-copy in
`save_file` be the preferred fix, or a clear error directing users to
`.clone()` / save-to-different-path?

**Comment by benchislett** [comment]
Is it possible to just have `save_file` write into a tempfile and then rename to the destination? That way it's not clobbering over the source file?

**Comment by benchislett** [comment]
Seems like problem still exists on main
